riscv/src/small_field/code_gen.rs

```diff
@@ -0,0 +1,1793 @@
+"# + "std::machines::small_field::memory::Memory regs(bit12, byte2);"
+        + r#"
+    // Get the value in register Y.
+    instr get_reg YL -> XH, XL link ~> (XH, XL) = regs.mload(0, YL, STEP);
+
+    // Set the value in register X to the value in register Y.
+    instr set_reg XL, YH, YL -> link ~> regs.mstore(0, XL, STEP, YH, YL);
+
+    // We still need these registers prover inputs.
+    reg query_arg_1_l;
+    reg query_arg_1_h;
+    reg query_arg_2_l;
+    reg query_arg_2_h;
+
+    // Witness columns used for intermediate values inside instructions.
+    col witness tmp1_h;
+    col witness tmp1_l;
+    col witness tmp2_h;
+    col witness tmp2_l;
+    col witness tmp3_h;
+    col witness tmp3_l;
+    col witness tmp4_h;
+    col witness tmp4_l;
+    col witness tmp5_h;
+    col witness tmp5_l;
+    col witness tmp6_h;
+    col witness tmp6_l;
+
+    link => byte2.check(tmp1_h);
+    link => byte2.check(tmp1_l);
+    link => byte2.check(tmp2_h);
+    link => byte2.check(tmp2_l);
+    link => byte2.check(tmp3_h);
+    link => byte2.check(tmp3_l);
+    link => byte2.check(tmp4_h);
+    link => byte2.check(tmp4_l);
+    link => byte2.check(tmp5_h);
+    link => byte2.check(tmp5_l);
+    link => byte2.check(tmp6_h);
+    link => byte2.check(tmp6_l);
+
+    // We need to add these inline instead of using std::utils::is_zero
+    // because when XX is not constrained, witgen will try to set XX,
+    // XX_inv and XXIsZero to zero, which fails this constraint.
+    // Therefore, we have to activate the following constraint whenever XXIsZero is used:
+    // XXIsZero = 1 - XX * XX_inv
+    col witness XX, XX_inv, XXIsZero;
+    std::utils::force_bool(XXIsZero);
+    XXIsZero * XX = 0;
+
+    // ============== control-flow instructions ==============
+
+    // Load the value of label `l` into register X.
+    // We restrict the address to 24 bits to avoid overflows.
+    instr load_label XL, l: label
+        link ~> regs.mstore(0, XL, STEP, tmp1_h, tmp1_l)
+        link => byte.check(tmp1_h)
+    {
+        tmp1_h * 2**16 + tmp1_l = l
+    }
+
+    // Jump to `l` and store the return program counter in register W.
+    // We restrict the address to 24 bits to avoid overflows.
+    instr jump l: label, WL
+        link ~> regs.mstore(0, WL, STEP, tmp1_h, tmp1_l)
+        link => byte.check(tmp1_h)
+    {
+        pc + 1 = (tmp1_h * 2**16) + tmp1_l,
+        pc' = l
+    }
+    
+    col witness tmp_fe;
+
+    // Jump to the address in register X and store the return program counter in register W.
+    instr jump_dyn XL, WL
+        link ~> (tmp1_h, tmp1_l) = regs.mload(0, XL, STEP)
+        link ~> regs.mstore(0, WL, STEP, tmp2_h, tmp2_l)
+        // pc is capped at 24 bits, so for this instruction 
+        // we restrict the higher limbs to 1 byte
+        link => byte.check(tmp1_h)
+        link => byte.check(tmp2_h)
+    {
+        // We need a temporary col here because the compiler doesn't like
```

Comment:
We could also fix it. The only reason we have this is because we are excessively trying to reduce the multplicative degree.

Reply:
Here is the fix: https://github.com/powdr-labs/powdr/pull/1893

Reply:
Thanks!